src/steps/Microsoft.DotNet.UpgradeAssistant.Steps.Packages/PackageCollection{T}.cs

```diff
@@ -0,0 +1,59 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+using Microsoft.DotNet.UpgradeAssistant.Packages;
+
+namespace Microsoft.DotNet.UpgradeAssistant.Steps.Packages
+{
+    internal class PackageCollection<T> : IDependencyCollection<T>
+    {
+        private readonly IEnumerable<T> _initial;
+        private readonly Action<BuildBreakRisk> _setRisk;
+
+        public PackageCollection(IEnumerable<T> initial, Action<BuildBreakRisk> setRisk)
+        {
+            _initial = initial;
+            _setRisk = setRisk;
+        }
+
+        public HashSet<T> Additions { get; } = new HashSet<T>();
+
+        public HashSet<T> Deletions { get; } = new HashSet<T>();
+
+        bool IDependencyCollection<T>.Add(T item, BuildBreakRisk risk)
+        {
+            _setRisk(risk);
+            return Additions.Add(item);
```

Comment:
Maybe make this conditional on `_initial` and `Additions` not containing `item`?